### PR TITLE
7483 app action interface for bdf2 stepper

### PR DESCRIPTION
**Pull request by asteyer**

### CODE REVIEW

File: packages/tempus/src/Tempus_StepperBDF2_impl.hpp [hunk 1]
```diff
@@ -64,7 +68,35 @@ StepperBDF2<Scalar>::StepperBDF2(
...
+  bool zeroInitialGuess,
+  const Teuchos::RCP<StepperBDF2AppAction<Scalar> >& stepperBDF2AppAction)
+{
+  this->setStepperType(        "BDF2");
+  this->setUseFSAL(            useFSAL);
+  this->setICConsistency(      ICConsistency);
+  this->setICConsistencyCheck( ICConsistencyCheck);
+  this->setZeroInitialGuess(   zeroInitialGuess);
+
+  this->setObserver();
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2_impl.hpp:90]: Need an #ifndef TEMPUS_HIDE_DEPRECATED_CODE around the setObserver.

File: packages/tempus/src/Tempus_StepperBDF2_impl.hpp [hunk 2]
```diff
@@ -198,6 +246,10 @@ void StepperBDF2<Scalar>::takeStep(
     //whether interpolated states are OK to use.
 
     //this->stepperObserver_->observeBeginTakeStep(solutionHistory, *this);
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2_impl.hpp:248]: Go ahead and remove the comment.

File: packages/tempus/src/Tempus_StepperBDF2_impl.hpp [hunk 3]
```diff
@@ -240,6 +300,8 @@ void StepperBDF2<Scalar>::takeStep(
     workingState->setOrder(getOrder());
     workingState->computeNorms(currentState);
     //this->stepperObserver_->observeEndTakeStep(solutionHistory, *this);
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2_impl.hpp:302]: Go ahead and remove comment.

File: packages/tempus/src/Tempus_StepperLeapfrog_decl.hpp [hunk 4]
```diff
@@ -12,7 +12,11 @@
 #include "Tempus_config.hpp"
 #include "Tempus_StepperExplicit.hpp"
 #include "Tempus_StepperObserverComposite.hpp"
-#include "Tempus_StepperLeapfrogObserver.hpp"
+#ifndef TEMPUS_HIDE_DEPRECATED_CODE
+  #include "Tempus_StepperLeapfrogObserver.hpp"
+#endif
+#include "Tempus_StepperLeapfrogAppAction.hpp"
+#include "Tempus_StepperLeapfrogAppActionComposite.hpp"
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperLeapfrog_decl.hpp:19]: Let's go ahead and remove the *Composite* code.  The Stepper should only need to know about the AppAction, and not whether if it is an AppAction composite.

File: packages/tempus/src/Tempus_StepperLeapfrog_impl.hpp [hunk 5]
```diff
@@ -42,15 +45,41 @@ StepperLeapfrog<Scalar>::StepperLeapfrog(
   this->setICConsistencyCheck( ICConsistencyCheck);
 
   this->setObserver(obs);
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperLeapfrog_impl.hpp:47]: Need an #ifndef TEMPUS_HIDE_DEPRECATED_CODE around the setObserver.

Comment (ccober6) [reply packages/tempus/src/Tempus_StepperLeapfrog_impl.hpp:47]: Oh sorry, not here but below.

File: packages/tempus/src/Tempus_StepperLeapfrog_impl.hpp [hunk 6]
```diff
@@ -42,15 +45,41 @@ StepperLeapfrog<Scalar>::StepperLeapfrog(
...
+  std::string ICConsistency,
+  bool ICConsistencyCheck,
+  const Teuchos::RCP<StepperLeapfrogAppAction<Scalar> >& stepperLFAppAction)
+  {
+    this->setStepperType(        "Leapfrog");
+    this->setUseFSAL(            useFSAL);
+    this->setICConsistency(      ICConsistency);
+    this->setICConsistencyCheck( ICConsistencyCheck);
 
+    this->setObserver();
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperLeapfrog_impl.hpp:71]: Need an #ifndef TEMPUS_HIDE_DEPRECATED_CODE around the setObserver.

File: packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  /// Observe BDF2 Stepper at end of takeStep.
+  virtual void observe(
+    Teuchos::RCP<const SolutionHistory<Scalar> > /* sh */,
+    Teuchos::RCP<const StepperBDF2<Scalar> > /* stepper */,
+    const typename StepperBDF2AppAction<Scalar>::ACTION_LOCATION actLoc)
+  {
+    switch(actLoc) {
+      case StepperBDF2AppAction<Scalar>::BEFORE_SOLVE:
+      case StepperBDF2AppAction<Scalar>::AFTER_SOLVE:
```
Comment (ccober6) [inline packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp:44]: Missing BEGIN_STEP and End_STEP?

**Review by ccober6 — CHANGES_REQUESTED** [review]
Just the one change.

File: packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp (hunk 7 above)
Comment (asteyer) [reply packages/tempus/src/Tempus_StepperBDF2ObserverDefault.hpp:45]: Latest change fixes this.

**Review by ccober6 — APPROVED** [review]
Looks good!

**Pull request closed without merging**